hs/app/reach/Main.hs

```diff
@@ -2450,6 +2451,149 @@ whoami = command "whoami" $ info f fullDesc
+      map splitByEqualsSigns stringArray
+
+    isDeviceCodePair :: [Text] -> Bool
+    isDeviceCodePair pair = head pair == pack "device_code"
+
+    isUserCodePair :: [Text] -> Bool
+    isUserCodePair pair = head pair == pack "user_code"
+
+    isAccessTokenPair :: [Text] -> Bool
+    isAccessTokenPair pair = head pair == pack "access_token"
+
+    clientId :: String
+    clientId = "b0e24d4cc8251c6cd14c"
+
+    scope :: String
+    scope = "gist"
+
+    language :: String
+    language = "JavaScript"
+
+    typeForUploadJson :: String
+    typeForUploadJson = "application/javascript"
+
+    acceptHeader = "Accept"
+    authorizationHeader = "Authorization"
+
+    step1 :: App
+    step1 = do
+      let dataJson =
+            object
+              [ "client_id" .= clientId
+              , "scope" .= scope
+              ]
+      parsedRequest <- parseRequest "POST https://github.com/login/device/code"
+      let request = setRequestBodyJSON dataJson parsedRequest
+      response <- httpLBS request
+      let githubResponseString = getResponseBody response
+      let arrayOfArrayOfText = process githubResponseString
+      let deviceCodePair = head $ filter isDeviceCodePair arrayOfArrayOfText
+      let deviceCode = deviceCodePair !! 1
+      let userCodePair = head $ filter isUserCodePair arrayOfArrayOfText
+      let userCode = userCodePair !! 1
+      liftIO . T.putStrLn $ pack $ "Your user code is " ++ unpack userCode
+      liftIO . T.putStrLn $ pack "Please enter it at https://github.com/login/device"
+      liftIO . T.putStrLn $ pack ""
+      liftIO . T.putStrLn $ pack
+        "Type 'y' AFTER SUCCESSFUL AUTHORIZATION to upload index.mjs and index.rsh."
+      userEnteredCharacter <- liftIO getChar
+      if toUpper userEnteredCharacter == 'Y'
+        then do step2WithThe deviceCode
+        else do
+          liftIO . T.putStrLn $ pack ""
+          liftIO . T.putStrLn $
+            "No files uploaded; run reach support again to make another try."
+
+    grantType :: String
+    grantType = "urn:ietf:params:oauth:grant-type:device_code"
+    userAgentHeader = "user-agent"
+
+    step2WithThe :: DeviceCode -> ReaderT Env IO ()
+    step2WithThe deviceCode = do
+      let dataJson =
+            object
+              [ "client_id" .= clientId
+              , "device_code" .= deviceCode
+              , "grant_type" .= grantType
+              ]
+      parsedRequest <- parseRequest "POST https://github.com/login/oauth/access_token"
+      let request = setRequestBodyJSON dataJson parsedRequest
+      response <- httpLBS request
+      let githubResponseString = getResponseBody response
+      let arrayOfArrayOfText = process githubResponseString
+      -- @TODO: Error handling! reach: Prelude.head: empty list
+      -- What if the user ignores instructions and hits 'y'
+      -- without actually authenticating, first?
+      let accessTokenPair = head $ filter isAccessTokenPair arrayOfArrayOfText
```

Comment:
headMay

Reply:
Ah, that's so much nicer! I think [18c8e4b](https://github.com/reach-sh/reach-lang/pull/762/commits/18c8e4b9989ca85b14b1c248aa6260e0606e949f) addresses this. Thank you for the recommendation.